Add/feature#3 area crud

FranckZnt19 opened the pull request:
# Description
 - creation area module whit scaffold CRUD
 - creation api area only index and show
 - testing with rspec
 
 # Testing
![imagen](https://user-images.githubusercontent.com/55332332/113243953-3e5a5800-9271-11eb-9262-1af92098e447.png)

  
  ## Screenshots 
![imagen](https://user-images.githubusercontent.com/55332332/113243692-b5432100-9270-11eb-9a4b-2248fad605cb.png)

Commit:
create area

this commit includes the cration of the area modul crud a gem that was causing an error when migrating is moved to test gems

Commit:
Api area
for this commit i create a controller for take data of the module area and i modificate the file route

Commit:
Validation
add the validation in module area

Commit:
test spec part 1
modification in the area module for api

Commit:
test spec part2
modification in file request area module

Commit:
test spec part3
modification file in area module for api

ArielMD commented on alerta-covid/app/assets/stylesheets/areas.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Areas controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
you can remove this comments, if are not necessary

FranckZnt19 replied:
comments were removed from this file

ArielMD commented on alerta-covid/app/controllers/areas_controller.rb:
```diff
@@ -0,0 +1,69 @@
+  # DELETE /areas/1 or /areas/1.json
+  def destroy
+    @area.destroy
+    respond_to do |format|
```
you dont need to respond to Json here, sice we have the api area controller

FranckZnt19 replied:
Removed json responses for web module controller

ArielMD commented on alerta-covid/app/views/areas/_area.json.jbuilder:
```diff
@@ -0,0 +1,2 @@
+json.extract! area, :id, :institution_id, :name, :created_at, :updated_at
+json.url area_url(area, format: :json)
```
if you remove the respond to Json in web controller, this file can also be removed

FranckZnt19 replied:
the json response is needed for the module's api

Commit:
Remove comments
comments were moved that are not needed in this file

Commit:
Remove json
Removed json responses for web module controller

Commit:
Merge branch 'master' into add/feature#3-area-crud

mariormt17 reviewed (dismissed):
The same comments of feature#4 apply to this one.

Commit:
config database
change data in file databas.yml for params corrects

EduardoGHdez commented on alerta-covid/config/database.yml:
```diff
@@ -1,86 +1,22 @@
-# PostgreSQL. Versions 9.3 and up are supported.
-#
-# Install the pg driver:
-#   gem install pg
-# On macOS with Homebrew:
-#   gem install pg -- --with-pg-config=/usr/local/bin/pg_config
-# On macOS with MacPorts:
-#   gem install pg -- --with-pg-config=/opt/local/lib/postgresql84/bin/pg_config
-# On Windows:
-#   gem install pg
-#       Choose the win32 build.
-#       Install PostgreSQL and put its /bin directory on your path.
-#
-# Configure Using Gemfile
-# gem 'pg'
-#
```
I'd suggest not tracking the `database.yml` file to avoid changing the `development` config.
Instead of this, I'd move it to a template -> `config/database.yml.example`

That is a common configuration to allow any contributor making they preferred db-configuration 😄

EduardoGHdez replied:
In other words: 
- Instead of making changes to this file, move it to `config/database.yml.example` and add `config/database.yml` to the `gitignore`

FranckZnt19 replied:
Thanks for this suggestion, we will take it into account in a new issue.

EduardoGHdez commented on alerta-covid/config/database.yml:
```diff
@@ -1,86 +1,22 @@
-  username: alerta_covid
-  password: <%= ENV['ALERTA_COVID_DATABASE_PASSWORD'] %>
+  database: reloj_checador_production
+  username: registros
```
This seems wrong. I mean, the project isn't the reloj-checador, right?

FranckZnt19 replied:
this error was fixed in the last commit of this branch

EduardoGHdez commented on alerta-covid/db/migrate/20210326062347_create_areas.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateAreas < ActiveRecord::Migration[6.1]
+  def change
+    create_table :areas do |t|
+      t.references :institution, null: false, foreign_key: true
+      t.string :name
+
+      t.timestamps
+    end
+  end
+end
```
Is a good practice to keep migrations in a single commit. But it's not required

FranckZnt19 replied:
Thanks for the observation I will take it into account in the following commits

EduardoGHdez commented on alerta-covid/app/assets/stylesheets/scaffolds.scss:
```diff
@@ -0,0 +1,65 @@
+th {
+  padding-bottom: 5px; }
+
+td {
+  padding: 0 5px 7px; }
+
+div.field,
+div.actions {
+  margin-bottom: 10px; }
+
+#notice {
+  color: green; }
+
+.field_with_errors {
+  padding: 2px;
+  background-color: red;
+  display: table; }
+
+#error_explanation {
+  width: 450px;
+  border: 2px solid red;
+  padding: 7px 7px 0;
+  margin-bottom: 20px;
+  background-color: #f0f0f0; }
+
+#error_explanation h2 {
+  text-align: left;
+  font-weight: bold;
+  padding: 5px 5px 5px 15px;
+  font-size: 12px;
+  margin: -7px -7px 0;
+  background-color: #c00;
+  color: #fff; }
+
+#error_explanation ul li {
+  font-size: 12px;
+  list-style: square; }
+
+label {
+  display: block; }
```
I know this was added by the scaffold command, but do we really want these style definitions?

FranckZnt19 replied:
for the front we will make separate issues where we will modify the style definitions for each of the project modules.

EduardoGHdez commented on alerta-covid/app/controllers/api/v1/areas_controller.rb:
```diff
@@ -0,0 +1,21 @@
+class Api::V1::AreasController < Api::V1::ApiController
+  before_action :set_area, only: %i[ show ]
+  skip_before_action :verify_authenticity_token
+
+  # GET /areas or /areas.json
+  def index
+    @areas = Area.all
+    render json: @areas, status: :ok
+  end
+
+  # GET /areas/1 or /areas/1.json
+  def show
+    render json: @areas, status: :ok
+  end
+
+  private
+    # Use callbacks to share common setup or constraints between actions.
+    def set_area
+      @area = Area.find(params[:id])
+    end
+end
```
Would be nice to add some tests to ensure this is behaving as expected 👍🏽

FranckZnt19 replied:
Thanks! the tests are in the following commits [ test spec part 1](https://github.com/BC-NOV-20-Proyectos/alerta-covid-web-team01/pull/13/commits/6820cd8317c2d624dd3193ed07be38dfbfd37151) [ test spec part2](https://github.com/BC-NOV-20-Proyectos/alerta-covid-web-team01/pull/13/commits/6a68bcb443badd8caea6a43616319dae0962bfc8) [ test spec part3](https://github.com/BC-NOV-20-Proyectos/alerta-covid-web-team01/pull/13/commits/526b3826aea051503f9306d3eb94acae37f92b8d)

EduardoGHdez commented on alerta-covid/spec/requests/api/v1/areas_spec.rb:
```diff
@@ -0,0 +1,29 @@
+
+  describe "GET /Api/V1/index" do
+    it "renders a successful response" do
+      get "Api/V1/areas"
```
Did you tried with `api/v1/areas`? 
I've never seen using capitals in routes.

I think the router does support it, but let's try to use conventional paths

FranckZnt19 replied:
the mentioned routes were modified in the next commit [test spec part2](https://github.com/BC-NOV-20-Proyectos/alerta-covid-web-team01/pull/13/commits/6a68bcb443badd8caea6a43616319dae0962bfc8)

EduardoGHdez commented on alerta-covid/spec/requests/api/v1/areas_spec.rb:
```diff
@@ -0,0 +1,29 @@
+ require 'rails_helper'
+
+RSpec.describe "/areas", type: :request do
```
🔥 Thanks for adding specs to this

FranckZnt19 replied:
Thank you very much for your comments!

EduardoGHdez commented on alerta-covid/spec/requests/api/v1/areas_spec.rb:
```diff
@@ -0,0 +1,29 @@
+ require 'rails_helper'
+
+RSpec.describe "/areas", type: :request do
+  let(:valid_attributes) {
```
This can be a one-liner

EduardoGHdez replied:
Also, I'd add space between brackets when using them 😄 
For example: 
```suggestion
  let(:valid_attributes) { { name: nil } }
```

FranckZnt19 replied:
ready, now commits based on these observations

EduardoGHdez commented on alerta-covid/spec/requests/api/v1/areas_spec.rb:
```diff
@@ -10,19 +10,20 @@
   }
 
   before(:each) do
+    FactoryBot.create(:institution, id:1, name: "institution")
```
You shouldn't specify the `id`

Any particular reason for doing it?

FranckZnt19 replied:
ready, now commits based on these observations

EduardoGHdez reviewed (commented):
Feedback
---

Regarding Git, I know it's hard at the beginning, but would recommend improving the git-history and try to create commits that actually do add value to the project. For example:
- Tests should be added when changing the behavior of any piece of code. In this case, I'd include the tests when adding a new controller or modifying one, and not creating a new commit only for specs
- There is no need to mention "the commit" in the commit message. It's enough just describing what we are trying to accomplish. For example, instead of
  - > create area
      this commit includes the cration of the area modul crud a gem that was causing an error when migrating is moved to test gems
  - I'd say something like this:
  - > Create Areas interface
    We want to have a UI to be able to create/update/read and delete an area
    For now, the only restriction to create an area is to include the `name`
  - A very specific purpose and let the other things were mentioned in the previous commit message would let them for a different commit.
   
 - I'd also suggest to avoid using Scaffold. It's really rare to use it in big projects. I think we had time for playing with it in the `reloj-checador` project, but I'd recommend to only add what you really need to get done your task 👍🏽 
  
Pls, let me know if what I mention make sense, and if you'd prefer chatting about this, my DM's are open to sync and have a call 😄

Commit:
syntax solution
Se modificó la sintaxis del código en los bloques de validación de las pruebas para tener un código mejor estructurado teniendo en cuenta la guía de estilo ruby

Commit:
update parameters
The fixed parameters of the objects created for the tests in rspec were eliminated and commands according to the ruby style guide were used to have a cleaner code
